### DISCUSSION TITLE
Smaller shapes / resizable shapes

### DISCUSSION THREAD

**Discussion by PaulKappmeyer** [hneemann/Digital#810] [Ideas]
The basic logic components have got the option named „Wide Shape“. Deselecting this option shrinks the icon of the gate.

I recognized that I use the smaller shapes most of the time – and I also saw people on YT doing the same.
Do you think setting the “Smaller Shape”-option as the standard option would be a good idea?

Or do you think it would be even possible to set the standard option in the general settings menu?

Also, I would like to ask how these shapes are generally produced? Because I think it would be even better to be able to set the width and height of the shapes (maybe in units of the grid-size) individually. Is this a possible/easy feature to implement?

**Comment by gixslayer** [comment]
I too find myself using the smaller shape most of the time. For that reason I think a global setting to control the default value of the *Wide Shape* attribute would be a nice addition. Usually I just set the value once and end up copy/pasting elements, so it's not a huge hassle either way.

The (IEEE) shapes are defined as static polygons, as seen [here](https://github.com/hneemann/Digital/blob/master/src/main/java/de/neemann/digital/draw/shapes/ieee/IEEEAndShape.java#L21) for the IEEE And gate shape. Technically I don't see why it wouldn't be possible to add *width* and *height* attributes, and generate the polygon based on them, rather than having two static variants as is now the case. This does add some complexity as you probably want to be careful and not needlessly re-create `Polygon` instances unless actually needed (I'm not sure how often the `drawIEEE` method is called), which is a bit trickier when the shapes are not static. A caching scheme could be the answer for that, but again this adds complexity.

Personally I don't really see the need to be able to control the shape to that extend. Even a generic 'scale' option feels overkill given the wide/non-wide shapes already available.

**Comment by hneemann** [comment]
The option of wider shapes exists only because the shapes used in English-speaking and in German-speaking countries for logical operations are not equally wide. If a circuit is created in one country and then opened in another country, the usual shape will still be displayed, but the width must be adjusted to preserve the circuit's function. Shapes that are slightly too wide or slightly too narrow will be displayed. If you create your own circuits, you don't have to worry about the width, because it is automatically adjusted to the usual width in your country.
One could have simply omitted this option. Then the user could not change it, and the shapes would have a somewhat unusual width when a circuit is opened from a country where other shapes are common.

**Reply by PaulKappmeyer** [reply]
Thanks for that explanation.

@gixslayer:
> Usually I just set the value once and end up copy/pasting elements, so it's not a huge hassle either way.

I use this as a kind of “workaround” too – but still: it seems like unnecessary work and it can get annoying when using a lot of different gates.


![Digital 1](https://user-images.githubusercontent.com/42696082/131501100-7614397f-61a0-4c6f-a289-5a7cad144eb8.png)

I’ve attached a screenshot of the settings menu with the addition I’m thinking of. So, it’s just about the standard value when placing a new gate. (Sine I prefer to work with the smaller shapes, but others might want to use the wider shapes instead, I think this could be an enhancement?)

Sorry if I understood your explanation wrong.
